Source/Particles/Gather/FieldGather.H

```diff
@@ -531,4 +531,226 @@ void doGatherShapeN (const amrex::ParticleReal xp,
+    {
+        return ex_arr.contains(ii,jj,kk) ? ex_arr(ii,jj,kk) : 0.0_rt;
+    };
+    const auto ey_arr_zeropad = [ey_arr] (const int ii, const int jj, const int kk) noexcept
+    {
+        return ey_arr.contains(ii,jj,kk) ? ey_arr(ii,jj,kk) : 0.0_rt;
+    };
+    const auto ez_arr_zeropad = [ez_arr] (const int ii, const int jj, const int kk) noexcept
+    {
+        return ez_arr.contains(ii,jj,kk) ? ez_arr(ii,jj,kk) : 0.0_rt;
+    };
+    const auto bx_arr_zeropad = [bx_arr] (const int ii, const int jj, const int kk) noexcept
+    {
+        return bx_arr.contains(ii,jj,kk) ? bx_arr(ii,jj,kk) : 0.0_rt;
+    };
+    const auto by_arr_zeropad = [by_arr] (const int ii, const int jj, const int kk) noexcept
+    {
+        return by_arr.contains(ii,jj,kk) ? by_arr(ii,jj,kk) : 0.0_rt;
+    };
+    const auto bz_arr_zeropad = [bz_arr] (const int ii, const int jj, const int kk) noexcept
+    {
+        return bz_arr.contains(ii,jj,kk) ? bz_arr(ii,jj,kk) : 0.0_rt;
+    };
+
+    const amrex::Real x = (xp - xyz_min[0]) / dxyz[0];
+#if (AMREX_SPACEDIM == 3)
+    const amrex::Real y = (yp - xyz_min[1]) / dxyz[1];
+#endif
+    const amrex::Real z = (zp - xyz_min[2]) / dxyz[2];
+
+    // Compute largest nearest integer index
+    const int ip = static_cast<int>(std::ceil(x));
+#if   (AMREX_SPACEDIM == 2)
+    const int jp = static_cast<int>(std::ceil(z));
+    const int kp = 0;
+#elif (AMREX_SPACEDIM == 3)
+    const int jp = static_cast<int>(std::ceil(y));
+    const int kp = static_cast<int>(std::ceil(z));
+#endif
+
+    const int noi = nox;
+#if   (AMREX_SPACEDIM == 2)
+    const int noj = noz;
+#elif (AMREX_SPACEDIM == 3)
+    const int noj = noy;
+    const int nok = noz;
+#endif
+
+    // Min and max for interpolation loop along i
+    const int imin = ip - noi/2;
+    const int imax = ip + noi/2 - 1;
+
+    // Min and max for interpolation loop along j
+    const int jmin = jp - noj/2;
+    const int jmax = jp + noj/2 - 1;
+
+    // Min and max for interpolation loop along k
+#if   (AMREX_SPACEDIM == 2)
+    // kp = 0 always
+    const int kmin = kp;
+    const int kmax = kp;
+#elif (AMREX_SPACEDIM == 3)
+    const int kmin = kp - nok/2;
+    const int kmax = kp + nok/2 - 1;
+#endif
+
+    // Number of interpolation points
+    const int ni = imax - imin;
+    const int nj = jmax - jmin;
+    const int nk = kmax - kmin;
+
+    // Compute difference r in each direction
+    // (r always negative, by construction)
+    const amrex::Real ri = x - ip;
+#if   (AMREX_SPACEDIM == 2)
+    const amrex::Real rj = z - jp;
+#elif (AMREX_SPACEDIM == 3)
+    const amrex::Real rj = y - jp;
+    const amrex::Real rk = z - kp;
+#endif
+
+    // Compute centering coefficients
+    amrex::Gpu::DeviceVector<amrex::Real> sci;
+    amrex::Gpu::DeviceVector<amrex::Real> scj;
+#if (AMREX_SPACEDIM == 3)
+    amrex::Gpu::DeviceVector<amrex::Real> sck;
+#endif
+
+    sci.resize(2*noi);
+    scj.resize(2*noj);
+#if (AMREX_SPACEDIM == 3)
+    sck.resize(2*nok);
+#endif
```

Comment:
@WeiqunZhang I guess I cannot do this from within a GPU kernel, correct? What would be an alternative, apart from declaring these vectors as members of the WarpX class and somehow passing them to the device kernel? I mean, what would be an alternative in order to keep the declaration of these vectors local to the device kernel? Thanks!

Reply:
Since `nox` etc. are not that big, maybe you can do
```
    amrex::Real sci[3]; // or amrex::GpuArray<amrex::Real, 3>
    amrex::Real scj[3];
    amrex::Real sck[3];
```
where `3` is the maximal possible `no?`.

Reply:
Yes, thanks. This is an option. At least to have something that builds on GPU for the moment, just to be able to run this code on GPU, before thinking of optimizations.